File: Graph(그래프)/백준_11725_트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

public class Main {
    static Queue<Integer> q = new LinkedList();
    static boolean[] check = new boolean[100001];
    static int[] parent_node = new int[100001];
    static ArrayList<Integer>[] edge = new ArrayList[100001];

    static void makeArrayList(ArrayList<Integer>[] arr_list) {
        for (int i = 0; i < arr_list.length; i++)
            arr_list[i] = new ArrayList();
    }

    static void BFS(int n) {
        q.add(n);
        check[n] = true;

        while (q.isEmpty() == false) {
            int f = q.peek();
            q.poll();
            for (int i : edge[f]) {
                if (check[i] == false) {
                    check[i] = true;
                    q.add(i);
                    parent_node[i] = f;
                }
            }
        }
    }

    public static void main(String[] args) {
        StringBuilder sb = new StringBuilder();
        Scanner s = new Scanner(System.in);
        makeArrayList(edge);
        int N = s.nextInt();
        int M = N - 1;
        while (M-- > 0) {
            int a, b;
            a = s.nextInt();
            b = s.nextInt();
            edge[a].add(b);
            edge[b].add(a);
        }

        BFS(1);

        for (int i = 2; i <= N; i++)
            sb.append(parent_node[i] + "\n");

        System.out.println(sb);

    }
}

File: Graph(그래프)/백준_11725_트리의 부모 찾기/트리의 부모 찾기.kt
import java.util.*

var check = Array<Boolean>(100001) { false }
var parent_node = Array<Int>(100001) { 0 }
var edge = Array<ArrayList<Int>>(100001) { ArrayList<Int>() }
var q = LinkedList<Int>()

fun BFS(n: Int) {
    q.add(n)
    check[n] = true

    while (q.isEmpty() == false) {
        var f = q.peek()
        q.poll()
        for (i in edge[f]) {
            if (check[i] == false) {
                check[i] = true
                parent_node[i] = f
                q.add(i)
            }
        }
    }
}

fun main(args: Array<String>) {
    var s = Scanner(System.`in`)
    var N = s.nextInt()
    var M = N - 1

    while (M-- > 0) {
        var a = 0
        var b = 0
        a = s.nextInt()
        b = s.nextInt()
        edge[a].add(b)
        edge[b].add(a)
    }

    BFS(1)

    for (i in 2..N) println(parent_node[i])
}
